jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp

```diff
@@ -0,0 +1,1815 @@
+
+        ecma_string_t *name_p = ecma_new_ecma_string_from_uint32 (length);
+        ecma_property_t *property_p = ecma_create_named_data_property (array_p, name_p,
+                                                                       true, true, true);
+        ecma_deref_ecma_string (name_p);
+
+        ecma_named_data_property_assign_value (array_p, property_p, value);
+        ecma_free_value (value, true);
+        length++;
+        parse_comma = true;
+      }
+
+      /*
+       * Parse error occured.
+       */
+      ecma_deref_object (array_p);
+      return ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+    }
+    default:
+    {
+      return ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+    }
+  }
+} /* ecma_builtin_json_parse_value */
+
+/**
+ * Abstract operation Walk defined in 15.12.2
+ *
+ * See also:
+ *          ECMA-262 v5, 15.12.2
+ *
+ * @return completion value
+ *         Returned value must be freed with ecma_free_completion_value.
+ */
+static ecma_completion_value_t
+ecma_builtin_json_walk (ecma_object_t *reviver_p, /**< reviver function */
+                        ecma_object_t *holder_p, /**< holder object */
+                        ecma_string_t *name_p) /** < property name */
+{
+  JERRY_ASSERT (reviver_p);
+  JERRY_ASSERT (holder_p);
+  JERRY_ASSERT (name_p);
+
+  ecma_completion_value_t completion_value = ecma_op_object_get (holder_p, name_p);
+
+  if (!ecma_is_completion_value_normal (completion_value))
+  {
+    return completion_value;
+  }
+
+  ecma_value_t value = ecma_copy_value (ecma_get_completion_value_value (completion_value), true);
+  ecma_free_completion_value (completion_value);
+  completion_value = ecma_make_empty_completion_value ();
+
+  if (ecma_is_value_object (value))
+  {
+    ecma_object_t *object_p = ecma_get_object_from_value (value);
+    ecma_property_t *properties_p = ecma_get_property_list (object_p);
+    uint32_t no_properties = 0;
+    ecma_property_t *property_p;
+
+    /*
+     * The following algorithm works with arrays and objects as well.
+     */
+
+    for (property_p = properties_p;
+         property_p != NULL;
+         property_p = ECMA_GET_POINTER (ecma_property_t, property_p->next_property_p))
+    {
+      /*
+       * All properties must be named properties, since we constructed them.
+       */
+      JERRY_ASSERT (property_p->type == ECMA_PROPERTY_NAMEDDATA
+                    || property_p->type == ECMA_PROPERTY_INTERNAL);
+
+      if (property_p->type == ECMA_PROPERTY_NAMEDDATA
+          && property_p->u.named_data_property.enumerable)
+      {
+        no_properties++;
+      }
+    }
+
+    if (no_properties > 0)
+    {
+      size_t buffer_size = sizeof (ecma_string_t*) * (no_properties);
+      ecma_string_t **property_names_p = (ecma_string_t**) mem_heap_alloc_block (buffer_size,
+                                                                                 MEM_HEAP_ALLOC_SHORT_TERM);
+
+      if (property_names_p == NULL)
```

Comment:
The case is possible only if `buffer_size` is zero, not in case of out-of-memory.